grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceMigrationService.groovy

```diff
@@ -0,0 +1,102 @@
+            // The amount field will be null for all pre-existing invoices since the field was not being used.
+            prepaymentInvoice.invoiceItems.find { it.amount == null }.each { InvoiceItem prepaymentInvoiceItem ->
+                prepaymentInvoiceItem.amount = computePrepaymentInvoiceItemAmount(prepaymentInvoiceItem)
+                prepaymentInvoiceItem.save(failOnError: true, flush: true)
+            }
+        }
+    }
+
+    // TODO: use transient field on order instead of this once that code is merged
+    private BigDecimal computePrepaymentInvoiceItemAmount(InvoiceItem prepaymentInvoiceItem) {
+        BigDecimal prepaymentPercent = (prepaymentInvoiceItem.order.paymentTerm?.prepaymentPercent ?: Constants.DEFAULT_PAYMENT_PERCENT) / 100
+
+        // If the invoice item is for an adjustment.
+        OrderAdjustment orderAdjustment = prepaymentInvoiceItem.orderAdjustment
+        if (orderAdjustment) {
+            return (orderAdjustment?.canceled ? 0 : orderAdjustment.totalAdjustments) * prepaymentPercent
+        }
+
+        // Else if the invoice item is for a canceled order item.
+        OrderItem orderItem = prepaymentInvoiceItem.orderItem
+        if (orderItem.orderItemStatusCode == OrderItemStatusCode.CANCELED) {
+            return 0
+        }
+
+        // Else the invoice item is for a non-canceled order item.
+        return (orderItem.quantity ?: 0) * (orderItem.unitPrice ?: 0.0) * prepaymentPercent
+    }
+
+    /**
+     * For STEP 2 of the migration.
+     *
+     * For every item in a given prepayment invoice, computes the inverse invoice item and insert it to the database
+     * by adding it to the given final invoice.
+     *
+     * This method assumes that because there is both a prepayment invoice and a final invoice, that the order
+     * associated with both invoices has been fully invoiced, and so there's no need to check each item of the
+     * prepayment invoice individually, they should ALL need to have inverse items created for them.
+     */
+    Invoice generateInverseInvoiceItems(Invoice prepaymentInvoice, Invoice finalInvoice) {
+        prepaymentInvoice.invoiceItems.each { InvoiceItem prepaymentInvoiceItem ->
+            InvoiceItem inverseItem = createInverseInvoiceItem(prepaymentInvoiceItem)
+            finalInvoice.addToInvoiceItems(inverseItem)
+        }
+        return finalInvoice.save(failOnError: true, flush: true)
+    }
+
+    private InvoiceItem createInverseInvoiceItem(InvoiceItem prepaymentInvoiceItem) {
+        // Because we already computed the amount field for prepayment invoice items in step 1, and because for
+        // existing data there will always be a one-to-one mapping of prepayment invoice item to final invoice item,
+        // all we need to do to create the inverse item is copy the prepayment invoice item and inverse the amount.
+        InvoiceItem inverseItem = new InvoiceItem()
```

Comment:
@awalkowiak I don't think this is the best way to clone the invoice item but I couldn't find much in the code about how we typically handle cloning domain objects (do we even do it?).

I'm thinking the best solution is a clone() method on the domain but it'd be nice if we just had some sort of default behaviour that we could wire into all our domain objects by default that will allow us to deep-clone things like relationships to other tables.

I didn't want to implement something like that before discussing it though (especially since GORM scares me) so I'm curious about your thoughts